Picked up items can now be added to offhand slots.

rasu3n opened the pull request:
## Introduction
<!-- Explain existing problems or why this pull request is necessary -->
Items that are currently picked up will not be added to the offhand slot.
### Relevant issues
<!-- List relevant issues here -->
<!--

* Fixes #1
* Fixes #2

-->

## Changes
### API changes
<!-- Any additions to the API that should be documented in release notes? -->

### Behavioural changes
<!-- Any change in how the server behaves, or its performance? -->

## Backwards compatibility
<!-- Any possible backwards incompatible changes? How are they solved, or how can they be solved? -->

## Follow-up
<!-- Suggest any actions to be done before/after merging this pull request -->
<!--

Requires translations:

| Name | Value in eng.ini |
| :--: | :---: |
| `foo.bar` | `Foo bar` |

-->

## Tests
<!--
Details should be provided of tests done. Simply saying "tested" or equivalent is not acceptable.

Attach scripts or actions to test this pull request, as well as the result
-->
(Sorry. #4359 has been applied to the server.)
https://user-images.githubusercontent.com/76860328/128990569-b5bd5a7f-03e4-40ae-ae8d-46ccbef1e318.mp4

Commit:
Picked up items can now be added to offhand slots.

Commit:
Oops

Commit:
Clean up imports

dktapps commented on src/entity/object/ItemEntity.php:
```diff
@@ -226,12 +226,18 @@ public function onCollideWithPlayer(Player $player) : void{
 			return;
 		}
 
+		$playerInventory ??= $player->getInventory();
```
Tangential to this PR, but it is kinda misleading to provide an inventory for pickup that won't actually accept the item (even if it is creative). I guess that points to a needed API change (inventories don't actually pick up items anyway, their holders (blocks or entities) do).

dktapps commented on src/entity/object/ItemEntity.php:
```diff
@@ -226,12 +226,18 @@ public function onCollideWithPlayer(Player $player) : void{
 		$item = $this->getItem();
-		$playerInventory = $player->getInventory();
+		$playerInventory = match(true){
+			$player->getOffHandInventory()->canAddItem($item) and !$player->getOffHandInventory()->isSlotEmpty(0) => $player->getOffHandInventory(),
```
It took me a couple of minutes to understand what this branch is doing.
A more self-explanatory check might be `$player->getOffHandInventory()->getItem(0)->canStackWith($item)`.

dktapps commented on src/entity/projectile/Arrow.php:
```diff
@@ -175,11 +175,17 @@ public function onCollideWithPlayer(Player $player) : void{
-		$playerInventory = $player->getInventory();
-		if($player->hasFiniteResources() and !$playerInventory->canAddItem($item)){
+		$playerInventory = match(true){
+			$player->getOffHandInventory()->canAddItem($item) and !$player->getOffHandInventory()->isSlotEmpty(0) => $player->getOffHandInventory(),
```
See above comment.

Commit:
Item::canStackWith() is smart for offhand slot checking.

Commit:
???

rasu3n committed:
Replace InventoryPickup*Event with EntityItemPickupEvent (#4384)

The rationale here is that inventories don't actually pick items up - their holders do.
It's especially misleading to say that an inventory is picking up an item in creative mode when the picked-up item can't actually be added to the target inventory in the first place.

This change allows a range of new functionality, such as:
- Allowing survival players to pick items up even when their inventories are full, similarly to creative players
- Changing the destination inventory of collected items (e.g. items could be redirected to the offhand or ender chest inventory, while still allowing other plugins to understand what's happening)

As an added bonus, this obsoletes one more use case for Inventory->getHolder(), bringing us one step closer to removing the cyclic reference nightmare from inventories.

The choice of naming (EntityItemPickup, instead of EntityPickupItem) is to be consistent with other events, where the word order is SubjectObjectActionEvent.

rasu3n committed:
conflict resolution

dktapps committed:
Merge branch 'master' into offhand

dktapps commented:
For future reference, resolving conflicts doesn't involve copy pasting the code commits. See https://git-scm.com/docs/git-merge.
